staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go

```diff
@@ -29,51 +29,98 @@ import (
+		case "http":
+			portStr = "80"
+		default:
+			portStr = "443"
+		}
+	}
+	return svcScheme, svcName, portStr, nil
+}
+
+// resolveToService returns the service, scheme and the port string (can be a name, can be a numeric port).
+func resolveToService(services listersv1.ServiceLister, namespace, id string) (*v1.Service, string, string, error) {
+	svc, err := services.Services(namespace).Get(id)
+	if err != nil {
+		return nil, "", "", err
 	}
 
+	return svc, "https", "443", nil
+}
+
+// findServicePort finds the service port by name or numerically.
+func findServicePort(svc *v1.Service, portStr string) (*v1.ServicePort, error) {
 	// If a port *number* was specified, find the corresponding service port name
 	if portNum, err := strconv.ParseInt(portStr, 10, 64); err == nil {
-		svc, err := services.Services(namespace).Get(svcName)
-		if err != nil {
-			return nil, err
-		}
-		found := false
 		for _, svcPort := range svc.Spec.Ports {
 			if int64(svcPort.Port) == portNum {
-				// use the declared port's name
-				portStr = svcPort.Name
-				found = true
-				break
+				return &svcPort, nil
 			}
 		}
-		if !found {
-			return nil, errors.NewServiceUnavailable(fmt.Sprintf("no service port %d found for service %q", portNum, svcName))
+	} else {
+		for _, svcPort := range svc.Spec.Ports {
+			if svcPort.Name == portStr {
+				return &svcPort, nil
+			}
 		}
 	}
+	return nil, errors.NewServiceUnavailable(fmt.Sprintf("no service port %q found for service %q", portStr, svc.Name))
+}
 
-	eps, err := endpoints.Endpoints(namespace).Get(svcName)
+// ResourceLocation returns a URL to which one can send traffic for the specified service.
+func ResolveEndpoint(services listersv1.ServiceLister, endpoints listersv1.EndpointsLister, namespace, id string) (*url.URL, error) {
+	svc, svcScheme, portStr, err := resolveToService(services, namespace, id)
+	if err != nil {
+		return nil, err
+	}
+
+	svcPort, err := findServicePort(svc, portStr)
+	if err != nil {
+		return nil, err
+	}
+
+	switch {
+	case svc.Spec.Type == v1.ServiceTypeClusterIP, svc.Spec.Type == v1.ServiceTypeLoadBalancer, svc.Spec.Type == v1.ServiceTypeNodePort:
+		// these are fine
+	default:
+		return nil, fmt.Errorf("unsupported service type %q", svc.Spec.Type)
+	}
+
+	eps, err := endpoints.Endpoints(namespace).Get(svc.Name)
 	if err != nil {
 		return nil, err
 	}
 	if len(eps.Subsets) == 0 {
-		return nil, errors.NewServiceUnavailable(fmt.Sprintf("no endpoints available for service %q", svcName))
+		return nil, errors.NewServiceUnavailable(fmt.Sprintf("no endpoints available for service %q", svc.Name))
 	}
+
 	// Pick a random Subset to start searching from.
 	ssSeed := rand.Intn(len(eps.Subsets))
+
 	// Find a Subset that has the port.
 	for ssi := 0; ssi < len(eps.Subsets); ssi++ {
 		ss := &eps.Subsets[(ssSeed+ssi)%len(eps.Subsets)]
 		if len(ss.Addresses) == 0 {
 			continue
 		}
 		for i := range ss.Ports {
-			if ss.Ports[i].Name == portStr {
+			if ss.Ports[i].Name == svcPort.Name {
```

Comment:
Not for now but there is something odd about the case where we are asked for a port number match, find the first matching port number and the pick all the routes which match that routes port name. If ports are consistently names there won't be an issue but if they aren't the behavior is going to be very hard to debug.